File: app/src/main/java/hpn332/cb/model/adapter/AdapterListTask.java
package hpn332.cb.model.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;

import hpn332.cb.R;
import hpn332.cb.ui.fragment.ListFragmentTask;
import hpn332.cb.utils.List;
import hpn332.cb.utils.Type;
import hpn332.cb.utils.Utils;
import hpn332.cb.model.database.ProviderHelper;
import hpn332.cb.model.stucture.TaskStructure;

public class AdapterListTask extends RecyclerView.Adapter<AdapterListTask.ItemHolder> {

	private static final String TAG = "AdapterListTask";

	private LayoutInflater                  inflater;
	private ArrayList<TaskStructure>        arrayList;
	private int                             step;
	private ListFragmentTask.OnStepFragment onStepFragment;

	public AdapterListTask(
			Context context, ArrayList<TaskStructure> arrayList, int step,
			ListFragmentTask.OnStepFragment onStepFragment) {

		this.onStepFragment = onStepFragment;
		this.arrayList = arrayList;
		inflater = LayoutInflater.from(context);
		this.step = step;

		Log.d(TAG, "AdapterListTask: " + step);
	}

	@Override
	public ItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View view = inflater.inflate(R.layout.row_item_task, parent, false);

		return new ItemHolder(view);
	}

	@Override
	public void onBindViewHolder(final ItemHolder holder, int position) {holder.onBind(position);}

	@Override
	public int getItemCount() {
		return arrayList.size();
	}

	class ItemHolder extends RecyclerView.ViewHolder implements View.OnLongClickListener {

		private TextView title, description, rank;
		private LinearLayout tagLayout;
		private ImageView    nextStep;
		private int          position;

		ItemHolder(View view) {
			super(view);

			title = view.findViewById(R.id.title_textView);
			description = view.findViewById(R.id.description_textView);
			rank = view.findViewById(R.id.rank_textView);

			nextStep = view.findViewById(R.id.next_step_imageView);

			tagLayout = view.findViewById(R.id.show_tag_layout);

			view.setOnLongClickListener(this);
		}

		void onBind(final int position) {

			this.position = position;

			title.setText(arrayList.get(position).getTitle());
			description.setText(arrayList.get(position).getDesc());
			rank.setText(String.valueOf(arrayList.get(position).getRank()));

			if (arrayList.get(position).getStep() < 3) {

				nextStep.setVisibility(View.VISIBLE);

				nextStep.setOnClickListener(view -> {
					ProviderHelper.updateOneTaskToNextStep(
							inflater.getContext(),
							arrayList.get(position).getId(),
							(arrayList.get(position).getStep() + 1));

					onStepFragment.onClickNext();
				});
			}


			if (!arrayList.get(position).getTag().isEmpty()) addTagColor();
		}

		private void addTagColor() {

			tagLayout.setVisibility(View.VISIBLE);

			ProviderHelper.queryListTag(inflater.getContext(), List.L_TAGS);

			String[] tags = arrayList.get(position).getTag().split("`");

			View[] view = new View[tags.length];

			int j = 0;
			for (String s : tags) {

				for (int i = 0; i < List.L_TAGS.size(); i++) {

					if (s.equals(String.valueOf(List.L_TAGS.get(i).getId()))) {
						view[j] = inflater.inflate(R.layout.view_tag_color, tagLayout, false);

						view[j].findViewById(R.id.view_tag_color).setBackgroundColor(
								List.L_TAGS.get(i).getColor());

						tagLayout.addView(view[j]);
					}
				}
				j++;
			}
		}

		@Override
		public boolean onLongClick(View view) {

			Log.d(TAG, "onLongClick: position : " + position);

			Utils.goTo(inflater.getContext(), Type.EDIT_TASK, position, step);

			return true;
		}
	}
}

File: app/src/main/java/hpn332/cb/view/task/AdapterListTask.kt
package hpn332.cb.view.task

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import java.util.ArrayList

import hpn332.cb.R
import hpn332.cb.utils.List
import hpn332.cb.utils.Type
import hpn332.cb.utils.Utils
import hpn332.cb.model.stucture.Task
import hpn332.cb.view.task.AListTask.Companion.vm
import kotlinx.android.synthetic.main.row_item_task.view.*

class AdapterListTask(
    context: Context,
    private val arrayList: ArrayList<Task>,
    private val step: Int) : RecyclerView.Adapter<AdapterListTask.ItemHolder>() {

	private val inflater: LayoutInflater = LayoutInflater.from(context)

    private val TAG = "AdapterListTask"

    init {
        Log.d(TAG, "AdapterListTask: " + step)
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemHolder {
		val view = inflater.inflate(R.layout.row_item_task, parent, false)

		return ItemHolder(view)
	}

	override fun onBindViewHolder(holder: ItemHolder, position: Int) = holder.onBind(position)


	override fun getItemCount(): Int = arrayList.size

	inner class ItemHolder(view: View) : RecyclerView.ViewHolder(view), View.OnLongClickListener {

		private val title: TextView = view.title_textView
        private val description: TextView = view.description_textView
        private val rank: TextView = view.rank_textView
        private val tagLayout: LinearLayout = view.show_tag_layout
        private val nextStep: ImageView = view.next_step_imageView

        init {
            view.setOnLongClickListener(this)
		}

		fun onBind(position: Int) {

			title.text = arrayList[position].title
			description.text = arrayList[position].desc
			rank.text = arrayList[position].rank.toString()

			if (arrayList[position].step < 3) {

				nextStep.visibility = View.VISIBLE

				nextStep.setOnClickListener {
//                    ProviderHelper.updateOneTaskToNextStep(
//							arrayList[position].id,
//							arrayList[position].step + 1)

					vm.onClickNext()
				}
			}

			if (!arrayList[position].tag_id.isEmpty()) addTagColor()
		}

		private fun addTagColor() {

			tagLayout.visibility = View.VISIBLE

//			ProviderHelper.queryListTag(List.L_TAGS)

			val tags = arrayList[adapterPosition].tag_id.split("`".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

			val view = arrayOfNulls<View>(tags.size)

            for ((j, s) in tags.withIndex()) {

				for (i in 0 until List.L_TAGS.size) {

					if (s == List.L_TAGS[i].id.toString()) {
						view[j] = inflater.inflate(R.layout.view_tag_color, tagLayout, false)

						view[j]?.findViewById<View>(R.id.view_tag_color)?.setBackgroundColor(
								List.L_TAGS[i].color)

						tagLayout.addView(view[j])
					}
				}
            }

//            var j = 0
//            for (s in tags) {
//
//                for (i in 0 until List.L_TAGS.size) {
//
//                    if (s == List.L_TAGS[i].id.toString()) {
//                        view[j] = inflater.inflate(R.layout.view_tag_color, tagLayout, false)
//
//                        view[j]?.findViewById<View>(R.id.view_tag_color)?.setBackgroundColor(
//                            List.L_TAGS[i].color)
//
//                        tagLayout.addView(view[j])
//                    }
//                }
//                j++
//            }

		}

		override fun onLongClick(view: View): Boolean {

			Log.d(TAG, "onLongClick: position : " + adapterPosition)

			Utils.goTo(inflater.context, Type.EDIT_TASK, adapterPosition, step)

			return true
		}
	}

}
